R-package/tests/testthat/test_learning_to_rank.R

```diff
@@ -0,0 +1,118 @@
+context("Learning to rank")
+
+test_that("learning-to-rank with lgb.train() works as expected", {
+    data(agaricus.train, package = "lightgbm")
+    # just keep a few features,to generate an model with imperfect fit
+    train <- agaricus.train
+    train_data <- train$data[, 1:20]
+    dtrain <- lgb.Dataset(
+        train_data
+        , label = train$label
+        , group = c(rep(4L, 1500L), 513L)
+    )
+    ndcg_at <- "1,2,3"
+    eval_names <-  paste0("ndcg@", strsplit(ndcg_at, ",")[[1]])
+    params <- list(
+        objective = "lambdarank"
+        , metric = "ndcg"
+        , ndcg_at = ndcg_at
+        , metric_freq = 1L
+        , max_position = 3L
+        , label_gain = "0,1,3,7"
+    )
+    model <- lgb.train(
+        params = params
+        , data = dtrain
+        , nrounds = 10L
+    )
+    expect_true(lgb.is.Booster(model))
+
+    dumped_model <- jsonlite::fromJSON(
+        model$dump_model()
+    )
+    expect_equal(dumped_model[["objective"]], "lambdarank")
+    expect_equal(dumped_model[["max_feature_idx"]], ncol(train_data) - 1L)
+
+    # check that evaluation results make sense (0.0 < nDCG < 1.0)
+    eval_results <- model$eval_train()
+    expect_equal(length(eval_results), length(eval_names))
+    for (result in eval_results){
+        expect_true(result[["value"]] > 0.0 && result[["value"]] < 1.0)
+        expect_true(result[["higher_better"]])
+        expect_identical(result[["data_name"]], "training")
+    }
+    expect_identical(sapply(eval_results, function(x){x$name}), eval_names)
+})
+
+test_that("learning-to-rank with lgb.cv() works as expected", {
+    data(agaricus.train, package = "lightgbm")
+    # just keep a few features,to generate an model with imperfect fit
+    train <- agaricus.train
+    train_data <- train$data[, 1:20]
+    dtrain <- lgb.Dataset(
+        train_data
+        , label = train$label
+        , group = c(rep(4L, 1500L), 513L)
```

Comment:
Should there be 1501 groups? (1500 groups of length 4 and 1 group of length 513?)

Reply:
That was my intention, but it's just an artifact of copying other code from the examples you shared. I "one big group" of 513 is probably unnecessary...I'm going to drop this so we keep only the first 6000 observations and split them into 1500 groups of 4, that's cleaner.

Reply:
fixed in https://github.com/microsoft/LightGBM/pull/2573/commits/ead17bcc6ddb06fa506071e76ce0bba213101f8c

Reply:
fixed in  https://github.com/microsoft/LightGBM/pull/2573/commits/ee037c9a2f32415c7bd7e63a4d1c093687d110e8 (had to rebase to `master` and force push with other R PRs that were just merged)